Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -639,35 +639,43 @@ extension BrowserViewController: WKNavigationDelegate {
     }
     
     func webView(webView: WKWebView, didCommitNavigation navigation: WKNavigation!) {
-        urlBar.updateURL(webView.URL);
-        toolbar.updateBackStatus(webView.canGoBack)
-        toolbar.updateFowardStatus(webView.canGoForward)
-        showToolbars(animated: false)
+        if let tab = tabManager.selectedTab {
+            if tab.webView == webView {
+                urlBar.updateURL(tab.displayURL);
+                toolbar.updateBackStatus(webView.canGoBack)
+                toolbar.updateFowardStatus(webView.canGoForward)
+                showToolbars(animated: false)
+
+                if let url = tab.displayURL?.absoluteString {
+                    profile.bookmarks.isBookmarked(url, success: { bookmarked in
+                        self.toolbar.updateBookmarkStatus(bookmarked)
+                    }, failure: { err in
+                        println("Error getting bookmark status: \(err)")
+                    })
+                }
 
-        if let url = webView.URL?.absoluteString {
-            profile.bookmarks.isBookmarked(url, success: { bookmarked in
-                self.toolbar.updateBookmarkStatus(bookmarked)
-            }, failure: { err in
-                println("Error getting bookmark status: \(err)")
-            })
+                updateInContentHomePanel(tab.displayURL)
+            }
         }
-
-        updateInContentHomePanel(webView.URL)
     }
 
     func webView(webView: WKWebView, didFinishNavigation navigation: WKNavigation!) {
-        let notificationCenter = NSNotificationCenter.defaultCenter()
-        var info = [NSObject: AnyObject]()
-        info["url"] = webView.URL
-        info["title"] = webView.title
-        notificationCenter.postNotificationName("LocationChange", object: self, userInfo: info)
-
-        UIAccessibilityPostNotification(UIAccessibilityScreenChangedNotification, nil)
-        // must be followed by LayoutChanged, as ScreenChanged will make VoiceOver
-        // cursor land on the correct initial element, but if not followed by LayoutChanged,
-        // VoiceOver will sometimes be stuck on the element, not allowing user to move
-        // forward/backward. Strange, but LayoutChanged fixes that.
-        UIAccessibilityPostNotification(UIAccessibilityLayoutChangedNotification, nil)
+        if let tab = tabManager.selectedTab {
+            if tab.webView == webView {
```

Comment:
Hm, do we even want this check here? Observers might be interested in `LocationChange` events even if that event happens on a non-selected tab. In fact, the only `LocationChange` observer we have right now is the one that adds the history entry to the DB. We wouldn't want to exclude background tabs from history, right?